### DISCUSSION TITLE
get web vitals via an API request

### DISCUSSION THREAD

**Discussion by dominnya** [vercel/next.js#44821] [Help]
Hello everyone, I was wondering if it's somehow possible to get web vitals information using a single API request.

# Context

Next.js provides an ability to work with [web-vitals](https://nextjs.org/docs/advanced-features/measuring-performance) but only using a function invoked for each request made to the server. From that point I wanted to get web vitals without use of clients visiting website along with an ability to refresh those without making an API request from client-side + without use of external APIs.

# My progress on solving my issue

I was exploring google for an answer but it seems like there's nothing matching my question other than #28769 that later got removed #36912.

## web-vitals-reporter

[web-vitals-reporter](https://github.com/treosh/web-vitals-reporter) on the other hand won't help me as it is only limited to a one request per session whereas I want to retrieve data using a request whenever I want + I want vice-versa for requests:

> The [web-vitals](https://github.com/GoogleChrome/web-vitals) library is excellent at accurately measuring [Web Vitals](https://web.dev/vitals/). But it has no opinion on how to report metrics from a browser to your analytics. It may result in multiple API calls, session tracking, and lost metrics. The web-vitals-reporter makes [Web Vitals](https://github.com/GoogleChrome/web-vitals) collecting as simple as sending one POST request.

## cypress-web-vitals

Then I found [cypress-web-vitals](https://github.com/cmorten/cypress-web-vitals) - a not so perfect solution:

> 1. The url is visited with the HTML response intercepted and modified by Cypress to include the [web-vitals](https://github.com/GoogleChrome/web-vitals#from-a-cdn) module script and some code to record the web-vitals values.
> 2. Several elements (if exist) starting with the provided element (based on `firstInputSelector`) are then clicked in quick succession to simulate a user clicking on the page. Note: if choosing a custom element, don't pick something that will navigate away from the page otherwise the plugin will fail to capture the web-vitals metrics.
> 3. The audit then waits for the page load event to allow for the values of LCP and CLS to settle; which are subject to change as different parts of the page load into view.
> 4. Next the audit simulates a page visibility state change [which is required for the CLS web-vital to be reported](https://www.npmjs.com/package/web-vitals#basic-usage).
> 5. The audit then attempts to wait for any outstanding web-vitals to be reported for which thresholds have been provided.
> 6. Finally the web-vitals values are compared against the thresholds, logging successful results and throwing an error for any unsuccessful signals. Note: if the audit was unable to record a web-vital then it is logged, _but the test will not fail_.

^ The solution above is good but **not** solving my issue for me hence I'm looking for an elegant solution without use of cypress & in-production.

## Own solution based off cypress-web-vitals

I tried rewriting the code provided by the dependency above and ended up with

```ts
import config from "./config";
import http from "http";
import { JSDOM } from "jsdom";

export class Vitals {
  private http: typeof http;
  private hostname: string;
  private port: number;
  private vitals: {
    defaultThresholds: {
      lcp: number;
      fid: number;
      cls: number;
      fcp: number;
      ttfb: number;
    };
    web: {
      accessorKey: string;
      core: string;
      snippet: string;
    };
  }

  constructor({
    hostname = config.address,
    port = config.port,
  }: {
    hostname?: string;
    port?: number;
  }) {
    this.http = http; // my attempts on adding interception
    this.hostname = hostname;
    this.port = port;
    this.vitals = {
      defaultThresholds: {
        lcp: 2500,
        fid: 100,
        cls: 0.1,
        fcp: 1800,
        ttfb: 600
      },
      web: {
        accessorKey: "__web_vitals__",
        core: "... taken from https://github.com/cmorten/cypress-web-vitals/blob/main/src/constants.js#L24"
        snippet: ""
      }
    }

    this.vitals.web.snippet = "... taken from https://github.com/cmorten/cypress-web-vitals/blob/main/src/constants.js#L26 but without script tag";
  }

  // parse to a dom and include vitals code from
  // https://github.com/cmorten/cypress-web-vitals/blob/main/src/constants.js
  private injectVitals(html: string): Document {
    const document = new JSDOM(html).window.document;
    const head = document.querySelector("head");
    const script = document.createElement("script");
    script.type = "module";
    script.innerHTML = this.vitals.web.snippet;
    head.appendChild(script);

    return document;
  }

  // include web vitals in a text/html request
  public async fetchWithVitals({
    hostname = this.hostname,
    port = this.port
  }: {
    hostname?: string;
    port?: number;
  }): Promise<Document> {
    // tried to work with interceptors but it's not working
    const response = await fetch(`http://${hostname}:${port}`);
    const html = await response.text();
    return this.injectVitals(html);
  }

  // A method to get vitals from a request
  public async getVitals({
    html = this.fetchWithVitals({})
  }: {
    html?: Promise<Document>;
  }): Promise<any> {
    const document = await html;
    return new Promise((resolve) => {
      // return vitals when found
      document.addEventListener(this.vitals.web.accessorKey, () => {
        if (document.defaultView && document.defaultView[this.vitals.web.accessorKey].cls) {
          resolve(document.defaultView[this.vitals.web.accessorKey]);
        }
      });
    });
  }
}
```

More about what's going on in the code:

- I tried reimplementing a method from the dependency
- basically based off [this](https://github.com/cmorten/cypress-web-vitals/blob/main/src/vitalsCommandHandler.js#L46-L62)
- failed to get vitals due to not resolving promise in getVitals...
- failed to implement interception & continue from [here](https://github.com/cmorten/cypress-web-vitals/blob/main/src/visitWithWebVitalsSnippet.js#L3)
- tried to do a single fetch 🤣
- as the result I failed

Looking forward for a help with my mess, thanks a lot.

**Comment by dominnya** [comment]
To clarify I try my best to implement web-vitals without adding any external code to my next.js app (making an API request from a custom server)
